be/src/exprs/map_functions.cpp

```diff
@@ -355,4 +355,133 @@ StatusOr<ColumnPtr> MapFunctions::distinct_map_keys(FunctionContext* context, co
     return map;
 }
 
+static inline std::tuple<NullColumnPtr, Column*, Column*, const UInt32Column*> unpack_map_column(
+        const ColumnPtr& input) {
+    NullColumnPtr map_null = nullptr;
+    MapColumn* map_col = nullptr;
+
+    auto map = ColumnHelper::unpack_and_duplicate_const_column(input->size(), input);
+    if (map->is_nullable()) {
+        auto nullable = down_cast<NullableColumn*>(map.get());
+        map_col = down_cast<MapColumn*>(nullable->data_column().get());
+        map_null = NullColumn::create(*nullable->null_column());
+    } else {
+        map_null = NullColumn::create(input->size(), 0);
+        map_col = down_cast<MapColumn*>(map.get());
+    }
+
+    return {map_null, map_col->keys_column().get(), map_col->values_column().get(), &map_col->offsets()};
+}
+
+// return nullable map
+StatusOr<ColumnPtr> MapFunctions::map_concat(FunctionContext* context, const Columns& columns) {
+    DCHECK_GT(columns.size(), 0);
+    Columns not_null_columns;
+    auto chunk_size = columns[0]->size();
+
+    // remove null columns
+    for (auto& col : columns) {
+        if (col->only_null()) {
+            continue;
+        }
+        ColumnPtr src_column = col;
+        if (col->is_constant()) {
+            src_column = ColumnHelper::unpack_and_duplicate_const_column(chunk_size, col);
+        }
+        not_null_columns.push_back(src_column);
+    }
+    if (not_null_columns.empty()) {
+        return ColumnHelper::create_const_null_column(chunk_size);
+    } else if (not_null_columns.size() == 1) {
+        return ColumnHelper::cast_to_nullable_column(not_null_columns[0]->clone_shared());
+    }
+
+    ssize_t columns_num = not_null_columns.size();
+    std::unique_ptr<uint32_t[]> hash_values[columns_num];
```

Comment:
why not use vector of vector

Reply:
is uint32_t enough to represent the hash ?

Reply:
it is used to store the hash values from diffrerent keys, one array per map keys with various length. the inner array should get raw pointer to compute hash as L:409, learned in `array_distinct()`.